### PR TITLE
 A simple try for HKT

### LINKED ISSUE #1213: Allow classes to be parametric in other parametric classes
This is a proposal for allowing generics as type parameters.  It's currently possible to write specific examples of monads, but in order to write the interface that all monads satisfy, I propose writing

``` javascript
interface Monad<T<~>> {
  map<A, B>(f: (a: A) => B): T<A> => T<B>;
  lift<A>(a: A): T<A>;
  join<A>(tta: T<T<A>>): T<A>;
}
```

Similarly, it's possible to write specific examples of cartesian functors, but in order to write the interface that all cartesian functors satisfy, I propose writing

``` javascript
interface Cartesian<T<~>> {
  all<A>(a: Array<T<A>>): T<Array<A>>;
}
```

Parametric type parameters can take any number of arguments:

``` javascript
interface Foo<T<~,~>> {
  bar<A, B>(f: (a: A) => B): T<A, B>;
}
```

That is, when a type parameter is followed by a tilde and a natural arity, the type parameter should be allowed to be used as a generic type with the given arity in the rest of the declaration.

Just as is the case now, when implementing such an interface, the generic type parameters should be filled in:

``` javascript
class ArrayMonad<A> implements Monad<Array> {
  map<A, B>(f: (a:A) => B): Array<A> => Array<B> {
    return (arr: Array<A>) =>  arr.map(f);
  }
  lift<A>(a: A): Array<A> { return [a]; }
  join<A>(tta: Array<Array<A>>): Array<A> {
    return tta.reduce((prev, cur) => prev.concat(cur));
  }
}
```

In addition to directly allowing compositions of generic types in the arguments, I propose that typedefs also support defining generics in this way (see [issue 308](https://github.com/Microsoft/TypeScript/issues/308)):

``` javascript
typedef Maybe<Array<~>> Composite<~> ;
class Foo implements Monad<Composite<~>> { ... }
```

The arities of the definition and the alias must match for the typedef to be valid.

### PR DESCRIPTION
**Pull request by ShuiRuTian**
Fixes #1213

Far from complemention, just a try for now.

---

Two list, one is added ability, one is added error check. 

Ability
- [x] class/interface accept type constructor as type parameter and map them correctly.
- [x] function accept type constructor as type parameter and map them correctly.
- [x] extends specific type constructor
- [ ] type constructor could accept and check constraint
``` ts
// constraint might be a very complex issue.
interface G1<T extends number>{}
interface G2<T extends string>{}
interface H1<Container<_> extends G1<_>>{}    // Container<string> should report error?
interface H2<Container<_ extends number>>{}
interface H3<Container<_> extends G1<_>|G2<_> >{}
interface H4<Container<_ extends number> extends G1<_ extends boolean>|G2<_ extends boolean> >{}

// Is H4 complex enough? 
// .......No, it just begin. I could not even give a constraint.
// Accorfing to the BFN, this is allowed.
interface H5<K1<K2<K3<_>>>>{}
// Could someone tell me what the hell is this? How could it be used in Scala or Haskell? And how to add constraint fot it?
```
- [ ] default typeConstructor parameter
- [ ] distinguish type constructor and proper type in quick info.
- [ ] fix services
    - [ ] quick info
    - [ ] auto complete

// from communicity 
- [ ] type guard could change type arguments in type constructor. https://github.com/microsoft/TypeScript/issues/1213#issuecomment-689744510

Error
- [ ] type parameters and type arguments must match for type constructor
- [ ] infer type must match, this is an example in scala.
``` scala
object HelloWorld {
   def main(args: Array[String]) {
      var q: List[Int] = List(1, 2, 3, 4)
      var w: Array[Boolean] = Array(true)
      var t = getData(q,q)
     
   }
   def getData[T<:Int,U<:Int,C[X<:Int]<:List[_]](a:C[T],b:C[U]) = 1/2
  
   def getData2[T,U,C[X],B[Q[_]]](a:C[T],b:C[U],c:B[C]) = 1/2
}
```

---

Here are some simple thoughts about constraints.

Most from scala.

Convert `Type Constructor Instance` to `Type Lambda`
`List = [T] =>> List[T]`

Convert `apply constraint` to `Type Lambda`
`type T[X] >: L <: U` convert to `type T >: ([X] =>> L) <: ([X] =>> U)`

Define `<:` on `TypeLambda`:
`type TL1  =  [X >: L1 <: U1] =>> R1`
`type TL2  =  [X >: L2 <: U2] =>> R2`
`TL1 <: TL2` if
  - the type interval `L2..U2` is contained in the type interval `L1..U1` (i.e. `L1 <: L2` and `U2 <: U1`) and
  - `R1 <: R2`
The conditions are reasonable: assume we have `interface F[G[T<:U1] <: G1[T]]`, `G[Timpl]`, so `Timpl <: U1`,
then we replace `G[T]` with `Gimpl[U<:U2] <: G2[U]`, if the U is stricter(`U2<:U1`), Timpl might not meet `Timpl <: U2`;
And we use functions in `G1[T]`, if `G2[U] <: G1[T]` these functions might not exist.

If `A <: B`
`A` and `B` both have met the contraints.
TypeLambda[A] = TypeLambda[B] is valid if TypeLambda[+T];
TypeLambda[B] = TypeLambda[A] is valid if TypeLambda[-T];

### CODE REVIEW

**Comment by Kingwl** [comment]
Cool!

Could you fix the test cases(or lint or merge master) and we could pack this.

**Comment by ShuiRuTian** [comment]
Most case should be fixed, only one failed. 
For temp I change the baseline to pass it, but in the future it would be fixed.

**Comment by ShuiRuTian** [comment]
Alright, finally......
It is a little harder than estimated, however, it passed........

**Comment by Kingwl** [comment]
seems not work again....

**Comment by ShuiRuTian** [comment]
Oh, I see the cute little green check mark and would not push any more until it is packed this time

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by Kingwl** [comment]
So coooooooooooooooooooooooool!

**Comment by xiaoxiangmoe** [comment]
Excited!

**Comment by lupuszr** [comment]
@ShuiRuTian  Really excited about this feature. But if I am not mistaken it seems like it doesn't work for types only interfaces. So we can't create a sum type like type Maybe<A> = Just<A> | Nothing;

**Comment by ShuiRuTian** [comment]
@lupuszr 
Whoops, I did not think about it.
It would be very helpful if there are some provided examples(the code you wish works). Usually issues come when trying to give a complex example.

**Comment by lupuszr** [comment]
@typescript-bot pack this.

**Comment by ShuiRuTian** [comment]
lol, only TS team member could do that.

**Comment by lupuszr** [comment]
Sorry I hoped I have magic powers like that :D 

Anyway what I tested in the last playground was:

@ShuiRuTian Are you sure if that works ? 

In the playground this was my test:

`interface Just<A> {value: A, tag: 'Maybe/Just'}
interface Nothing { tag: 'Maybe/Nothing'}
type Maybe<A> = Just<A> | Nothing;

type Functor<F<_>> = {
    map: <A, B>(f: (a: A) => B) => ((fa: F<A>) => F<B>)
}

type Monad<F<_>> = {
    bind: <A, B>(f: (a: A) => F<B>) => ((fa: F<A>) => F<B>)
}

type L = Functor<Maybe>`

and got an error Generic type 'Maybe' requires 1 type argument(s).(2314)

the same happens for:

`type Reader<E, A> = (env: E) => A
type R = Functor<Reader>` as well

Is this handled by your last commit ?

**Comment by ShuiRuTian** [comment]
Not sure, would check when having time. My break time is nearly at an end.

**Comment by ShuiRuTian** [comment]
@lupuszr  Auh, it should work for the sample code now. But I do not fix error and the code is really ugly.....
If you could not wait to see the result, you could fork and build this branch, then change the Typescript what vscode is using.

File: src/compiler/types.ts [hunk 1]
```diff
@@ -5340,17 +5340,17 @@ namespace ts {
     }
 
     // Type parameters (TypeFlags.TypeParameter | TypeFlags.TypeConstructorDeclaration)
-    export interface TypeConstructorDeclaration extends TypeParameter{
+    export interface TypeConstructorPolymophsimDeclaration extends TypeParameter{
```
Comment (lupuszr) [inline src/compiler/types.ts:5343]: you have a typo (missing r)

**Comment by orta** [comment]
@typescript-bot pack this

**Comment by Kingwl** [comment]
@typescript-bot pack this.

**Comment by jeremiahrhall** [comment]
I've tried this out locally and it's really interesting. The tests pass and I'm able to use it in VSCode. Is there anything I can help with here? Testing?

**Comment by ShuiRuTian** [comment]
@jeremiahrhall Thanks a lot for the kind willing!

I have almost forgot this PR..... for many reasons:

1. Not pretty clear about the syntax and concept. Although this PR could basically do the work asked for in the issue, but it is only the most simple case. Have no many thoughts about the advanced situation:
    1. What is "higher" kind type. As I have seen, when people talking about HKT, in most cases, it is actually "two kind type". Then issues come, what is the use of more higher kind type, should we support them, if yes what is its syntax and error...
    2. A big problem not mentioned is the constraint. As you could see  in my description, how to add constraint is not that clear and might be a very difficult thing, especially if we want to support three and more kind type. let me explain the difficult a bit:
        - data struct. Could current data struct support constraint? If not, We might need change ten thousands of lines.
        - compatibility. Other functions provided by TS, like `infer` keyword, how to be compitated with HKT?

2. It would be a little awkward to mention, but some personal stuff block me, I hope all things would go well so that I could have more free time.

3. I have some other PRs ready for reivew, as far as I'm concerned, I would like to see them merged first, then focusing on the most hard and interesting one.

And, yes, of course, testing this would be really helpful, especially when it could fit all your need. I am curious whether it is a common partten without needing constraint. If so, perhaps we could argue a bit with TS Team!

**Comment by 2A5F** [comment]
@ShuiRuTian 
for what is "higher", constraint, infer, this can be as a reference 
#35816

**Comment by ShuiRuTian** [comment]
@2A5F 

Thanks for the kind willing.

Some parts of the link looks good. However, other parts are basic and lack details.

1. higher kinded type
*->*, this is generic, one kind.
(*->*)->*, this is two kind, and in most cases, this is the form when people give examples.
((*->*)->*)->*, this is three kind, and it is also HKT.
So, HKT syntax should be able to express about TEN kind and more, although this might be ridiculous.

2. constraint
And this part, an important question is whether one constraint meet another constraint.
``` ts
interface MyGeneric<T extends string>{}
const foo: MyGeneric<number>; // error, number is not string
```
And in HKT, this is much more complex
``` ts
interface MyGeneric1<T extends string>{}
interface MyGeneric2<T extends number>{}
interface MyFunctor<G<T extends string> extends MyGeneric1<T>>{}
const foo: MyFunctor<MyGeneric2>; // error, number is not string, and MyGeneric2 is not MyGeneric1
// Think about three kinded HKT, it would be complex.
// Look this:
// interface HKT3<Functor<Generic1<T extends number> extends Generic2<T>, Generic2<U exnteds V> exnteds Set<V>>, V extends string>
// Is HKT3 valid?
// If we have HKT3<FunctorImpl, string>, should this be valid?
```

The conclution is type parameter should be cotravariance, and return type should be covariance.

3. infer

I do not have many idea about this, but it must be difficult.


And here is one question I am not sure:
(*->*)->(*->*), this is indeed HKT, but in which condition, could we return a Type Constructor?

**Comment by 2A5F** [comment]
@ShuiRuTian 

HKT3 certainly cannot be valid

```ts
type Foo<V extends string, U extends V> = U
type Bar<V extends string, T extends number> = Foo<V, T>

// Type 'T' does not satisfy the constraint 'V'.
//   'V' could be instantiated with an arbitrary type which could be unrelated to 'T'.
//     Type 'number' is not assignable to type 'V'.
//       'V' could be instantiated with an arbitrary type which could be unrelated to 'number'.
```
V is passed in like a closure

[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nAPANShAHsCA7AJgZyn2ACcBLbAcwBooBVNTHAqZAPigF56BYAKFCQoAIQCGJFIyx5CxclVoAVKc0LYArgFsARhBIdu8JMiVt+QA)

---

```ts
interface MyGeneric1<T extends string>{}
interface MyGeneric2<T extends number>{}
interface MyFunctor<
  G extends for<T extends string> MyGeneric1<T>
>{}
const foo: MyFunctor<MyGeneric2>;
```
I think my syntax is more readable and easier to think  
`for<T extends string> MyGeneric1<T>`  
just like a function or type kind syntax 
`(T: string) => MyGeneric1<T>` ,  `* -> *`  

---

```ts
interface HKT3<
  Functor extends for<
     Generic1 extends for<T extends number> Generic2<T>, // T error
     Generic2 exnteds for<U exnteds V> Set<V>
  >, 
  V extends string
>
```
how write FunctorImpl ?
```ts
interface FunctorImpl<
  Generic1 extends for<T extends number> Generic2<T>, // T error
  Generic2 exnteds for<U exnteds string> Set<string>
> { }

// or

type G2<U exnteds string>= Set<string>
type G1<T extends number> = G2<T> // T error

interface FunctorImpl<
  Generic1 extends G2,
  Generic2  extends G1,
>{ }
```

---

```ts
for<U exnteds string> Set<string> <- for<U extends string | number> Set<'a'>
// just like
(u: string) => Set<string> <- (u: string | number) => Set<'a'>
```

> In other words, the `→` type constructor is contravariant in the parameter (input) type and covariant in the return (output) type.
[wiki](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science))

---

> but in which condition, could we return a Type Constructor?

If a type constructor is not applied

```ts
type Foo<T> = T
type Bar = Foo // Bar is type Bar<T> = T
```

**Comment by sandersn** [comment]
This experiment is pretty old, so I'm going to close it to reduce the number of open PRs.

**Comment by rokinsky** [comment]
@sandersn, your development lead approved the approach and asked for a draft implementation — https://github.com/microsoft/TypeScript/issues/1213#issuecomment-96854288.

The case with one type parameter is basically what we need in 95%, e.g.
```typescript
interface Monad<F<_>> {
  flatMap<A, B>(f: (a: A) => B): F<A> => F<B>;
}
```

I hope you have closed this PR to continue working on the implementation within the Microsoft team, as the HKT topic is enormously in demand — you can look at the related discussion. 

For a reference, you can take an advanced Scala compiler and their HKT implementation.

**Comment by aliakbarazizi** [comment]
@sandersn @lupuszr  any update on this?
this pull request exist for more than 3 years and related issue is open for almost 9 years
https://github.com/Microsoft/TypeScript/issues/1213

**Comment by Anoesj** [comment]
@sandersn A bit weird how this was just closed without any sort of feedback or status update.. Where does MS stand regarding HKT/HOT nowadays?

**Pull request closed without merging**